tests/test_requests.py

```diff
@@ -1386,6 +1401,64 @@ def test_manual_redirect_with_partial_body_read(self, httpbin):
         r3 = next(rg)
         assert not r3.is_redirect
 
+    def test_rewind_body(self):
+        data = io.BytesIO(b'the data')
+        s = requests.Session()
+        prep = requests.Request('GET', 'http://example.com', data=data).prepare()
+        prep.body.read()
+
+        # the data has all been read
+        assert prep.body.read() == b''
+
+        # rewind it back
+        s.rewind_body(prep)
+        assert prep.body.read() == b'the data'
+
+    def test_rewind_partially_read_body(self):
+        data = io.BytesIO(b'the data')
+        s = requests.Session()
+        data.read(4)  # read some data
+        prep = requests.Request('GET', 'http://example.com', data=data).prepare()
+        prep.body.read()
+
+        # the data has all been read
+        assert prep.body.read() == b''
+
+        # rewind it back
+        s.rewind_body(prep)
+        assert prep.body.read() == b'data'
+
+    def test_rewind_body_no_seek(self):
+        class BadFileObj:
+            def __init__(self, data):
+                self.data = data
+
+        data = BadFileObj('the data')
+        s = requests.Session()
+        prep = requests.Request('GET', 'http://example.com', data=data).prepare()
+
+        with pytest.raises(RedirectError) as e:
+            s.rewind_body(prep)
+
+        assert 'Unable to read request body' in str(e)
+
+    def test_rewind_body_failed_seek(self):
+        class BadFileObj:
+            def __init__(self, data):
+                self.data = data
+
+            def seek(self, pos):
+                raise OSError()
+
+        data = BadFileObj('the data')
+        s = requests.Session()
+        prep = requests.Request('GET', 'http://example.com', data=data).prepare()
+
+        with pytest.raises(RedirectError) as e:
+            s.rewind_body(prep)
+
+        assert 'error occured when rewinding request body' in str(e)
+
```

Comment:
Presumably we should have a test here for an object with a `seek()` method but no working `tell()` method?